x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/EnrichPolicyResolver.java

```diff
@@ -110,22 +114,37 @@ public static UnresolvedPolicy from(Enrich e) {
     /**
      * Resolves a set of enrich policies
      *
-     * @param enriches           the unresolved policies
+     * @param preAnalysis        to retrieve indices and enriches to resolve
+     * @param requestFilter      to resolve target clusters
      * @param executionInfo      the execution info
      * @param listener           notified with the enrich resolution
      */
-    public void resolvePolicies(List<Enrich> enriches, EsqlExecutionInfo executionInfo, ActionListener<EnrichResolution> listener) {
-        if (enriches.isEmpty()) {
+    public void resolvePolicies(
+        PreAnalyzer.PreAnalysis preAnalysis,
+        QueryBuilder requestFilter,
+        EsqlExecutionInfo executionInfo,
+        ActionListener<EnrichResolution> listener
+    ) {
+        if (preAnalysis.enriches.isEmpty()) {
             listener.onResponse(new EnrichResolution());
             return;
         }
 
-        doResolvePolicies(
-            new HashSet<>(executionInfo.getClusters().keySet()),
-            enriches.stream().map(EnrichPolicyResolver.UnresolvedPolicy::from).toList(),
-            executionInfo,
-            listener
-        );
+        doResolveRemotes(preAnalysis.indices, requestFilter, listener.delegateFailureAndWrap((l, remotes) -> {
+            doResolvePolicies(remotes, preAnalysis.enriches.stream().map(UnresolvedPolicy::from).toList(), executionInfo, l);
+        }));
+    }
+
+    private void doResolveRemotes(List<IndexPattern> indexPatterns, QueryBuilder requestFilter, ActionListener<Set<String>> listener) {
+        switch (indexPatterns.size()) {
+            case 0 -> listener.onResponse(Set.of());
+            case 1 -> indexResolver.resolveConcreteIndices(
```

Comment:
If the index pattern also contains a cluster/project prefix, will this still be correct? 

eg.

```
FROM remote1:idx1,remote2:idx2
| STATS max(a) by b
| ENRICH policy on a, b
```

The enrich can only be calculated on the coordinator (after the STATS reduction), but I guess the enrich index on the coordinator won't be resolved. I'm not sure it works at all _today_ tbh, maybe worth checking.

Reply:
For LOOKUP JOINs we have a specific [check](https://github.com/elastic/elasticsearch/pull/134127): we just don't allow it.

I'm checking ENRICH now, but I couldn't find similar validation for it in the codebase, so I guess it's being treated differently...

Reply:
I believe it is behaving the same way. It looks like we have a similar test for enrich:

https://github.com/elastic/elasticsearch/blob/37b28e4983ddadc046d45a5806b36d58d55ea41b/x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/CrossClusterEnrichIT.java#L398-L421

Reply:
@idegtiarenko I added a test in `main` for this case, see https://github.com/elastic/elasticsearch/pull/134199, and it seems to work fine.
I think it makes sense to run it on this PR and see if the results are the same.